modules/ml/src/main/java/org/apache/ignite/ml/tree/data/TreeDataIndex.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.tree.data;
+
+import org.apache.ignite.ml.tree.TreeFilter;
+
+/**
+ * Index for representing sorted dataset rows for each features.
+ * It may be reused while decision tree learning at several levels through filter method.
+ */
+public class TreeDataIndex {
+    /** Index containing IDs of rows as if they is sorted by feature values. */
+    private final int[][] index;
+    /** Original features table. */
+    private final double[][] features;
+    /** Original labels. */
+    private final double[] labels;
+
+    /**
+     * Constructs an instance of TreeDataIndex.
+     *
+     * @param features Features.
+     * @param labels Labels.
+     */
+    public TreeDataIndex(double[][] features, double[] labels) {
+        this.features = features;
+        this.labels = labels;
+
+        int rows = features.length;
+        int cols = features.length == 0 ? 0 : features[0].length;
+
+        double[][] featuresCopy = new double[rows][cols];
+        index = new int[rows][cols];
```

Comment:
I think it makes sense to use here `Arrays.copyOf` for `featuresCopy` and `Arrays.fill` for `index`. What do you think?
And do we actually need copy of features here?